cli/central-client/src/main/java/org/ballerinalang/central/client/Utils.java

```diff
@@ -419,9 +434,28 @@ private static String getPlatformFromBala(String balaName, String packageName, S
         return balaName.split(packageName + "-")[1].split("-" + version)[0];
     }
 
-    private static void extractBala(Path balaFilePath, Path balaFileDestPath) throws IOException {
+    private static void extractBala(Path balaFilePath, Path balaFileDestPath, String trueDigest, String packageName,
+            PrintStream outStream)
+            throws IOException, CentralClientException {
         Files.createDirectories(balaFileDestPath);
         URI zipURI = URI.create("jar:" + balaFilePath.toUri().toString());
+        byte[] hashInBytes = checkHash(balaFilePath.toString(), "SHA-256");
+
+        // If the hash value is not matching , throw an exception.
+        if (!("sha-256=" + bytesToHex(hashInBytes)).equals(trueDigest)) {
+            StringBuilder warning = new StringBuilder(
+                    String.format("*************************************************************%n" +
+                            "* WARNING: Some packages may not be from original source. *%n" +
+                            "*************************************************************%n%n" +
+                            "The following package was failed to verify the hash value %n" +
+                            packageName +
+                            "%n"));
+            if (outStream != null) {
+                outStream.println(warning.toString());
+            }
+
```

Comment:
Few suggestions
1. Can use Objects.equal to compare in L445 since it handles null values gracefully.
2. Might be better to extract the error message to a variable to increase readability
3. "sha-256=" Can be moved to a constant. I saw you have used that string in a couple of places